packages/ui-components/src/Input/BaseInput.tsx

```diff
@@ -0,0 +1,94 @@
+import React, {
+  ChangeEvent,
+  CSSProperties,
+  useCallback,
+  useEffect,
+  useMemo,
+  useState,
+} from "react";
+import classNames from "classnames/bind";
+import styles from "./styles.module.scss";
+
+export interface InputProps<T = string | number> {
+  type?: T extends string ? "text" : "number";
+  initialValue?: T;
+  value?: T;
+  onChange?: (value: T) => void;
+  className?: string;
+  style?: CSSProperties;
+  autoComplete?: string;
+  placeholder?: string;
+  name?: string;
+  disabled?: boolean;
+  invalid?: boolean;
+  prefix?: React.ReactNode;
+}
+
+const cx = classNames.bind(styles);
+
+export const BaseInput: React.FC<InputProps> = ({
+  name,
+  type = "text",
+  autoComplete = "off",
+  className,
+  style,
+  value: outerValue = "",
+  initialValue = "",
+  placeholder,
+  onChange,
+  disabled = false,
+  invalid = false,
+}) => {
+  const [value, setValue] = useState<string | number>(
+    outerValue || initialValue,
+  );
+  const [isDirty, setDirtyState] = useState<boolean>(false);
+
+  useEffect(() => {
+    if (!outerValue && !isDirty) {
+      return;
+    }
+    setDirtyState(true);
```

Comment:
@nathanstilwell , you're right it's useless duplication in `onChangeHandler`. Will remove it.
```
setDirtyState(true);
setValue(nextValue);
```

The current implementation is necessary to achieve the following conditions:
- `defaultValue` is shown only when `value` isn't set and user didn't change the text
- New `value` provided from outside of the component overrides `defaultValue` and the current value of Input
- Typed value by a user should not be immediately overridden by the outer `value` prop.

Basically it allows us to change input value in two ways: programmatically by changing `value` prop or by user input.
To manage two-way changes for input - it's necessary to maintain an internal state for value. So `value` props is not directly passed to <input> `value` attribute. Instead, change of `value` props (`outerValue`) is tracked in `useEffect`.

`isDirty` flag is required to handle one corner case - when `value` prop is set to empty string or user removes input text. We need to distinguish whether to show `initialValue` or not. After value is remove - it is dirty state and no needs to show initial value.

Reply:
Thank you for the description. I was concerned that I was missing something with the duplication.

I agree with your approach to handling a little internal state. I’m wonder how this will work with something like React Final Form, which is our proposed form solution in Cloud console (and I would like to use it to manage all forms if possible). My gut feeling is that it will be fine as the Field props will be passed down and we will probably still have cases where we want to us an Input without using a Form or Field from RFF.

Reply:
Well, it's a great example to add in Storybook - to check out how this our component work with RFF! 👍🏽